### PR TITLE
Use qsim for repeated sampling

### LINKED ISSUE #503: Use qsim for repeated sampling
Highlighted in #502.

`_sample_measure_results` currently calls Cirq for repeated sampling of non-noisy circuits. This is a serviceable solution, but is somewhat less performant than simulating and sampling all on the qsim side. This is magnified by #502, where noiseless circuits on more than 32 qubits are forced to follow the repeated-simulation path.

We should replace the Cirq-dependent behavior with a method that handles this entirely in qsim.

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Fixes #503.

`*_sample` and `*_sample_final` need to be different functions because they have incompatible behaviors:
- `*_sample` (the existing methods) sample from measurements in a single execution of a circuit.
- `*_sample_final` (the new methods) sample repeatedly from the final state of a circuit.

### CODE REVIEW

File: pybind_interface/pybind_main.cpp [hunk 1]
```diff
@@ -932,6 +946,16 @@ qtrajectory_simulate_moment_expectation_values(
 
 // Methods for sampling.
 
+std::vector<uint64_t> qsim_sample_final(
+    const py::dict &options, uint64_t num_samples) {
+  return SimulatorHelper::sample_final_state(options, false, num_samples);
+}
+
+std::vector<uint64_t> qtrajectory_sample_final(
```
Comment (sergeisakov) [inline pybind_interface/pybind_main.cpp:954]: This function is not used, right?

**Review by sergeisakov — APPROVED** [review]
LGTM. Does it make sense to add some non-trivial test?

File: pybind_interface/pybind_main.cpp (hunk 1 above)
Comment (95-martin-orion) [reply pybind_interface/pybind_main.cpp:954]: Correct - repeated sampling from the final state of a noisy circuit isn't something we currently do. As it stands it doesn't seem like there's much harm in keeping it, but I'm happy to remove it if there's a concern about confusing users or if it's weighing down the library

Comment (sergeisakov) [reply pybind_interface/pybind_main.cpp:954]: I think we can keep it.

**Comment by sergeisakov** [comment]
> > LGTM. Does it make sense to add some non-trivial test?
> 
> I considered adding tests for #502, but a 33-qubit circuit of any depth would be excessive for a unit test. Other than that, the existing Python tests cover repeated sampling pretty thoroughly - is there anything extra you see that needs testing?

It seems that the tests are sort of trivial. For instance, the circuit in `test_qsim_run_vs_cirq_run` (line 405 in qsimcirq_test.py) just swaps the |0> amplitude and the `run` method always measures the same bit string. I was thinking of, say, a 4- or 5-qubit circuit with less trivial output and running a few thousands repetitions and comparing the measurement histogram with the actual probabilities. I am not sure if it makes sense to add such a test.

**Comment by 95-martin-orion** [comment]
The pybind call is simple enough that I think it's safe to rely on the C++ tests for verifying sampling behavior. The only new complexity on the Python side is in the bit-management of converting `raw_results` to `full_results` - I don't think a circuit with multiple possible sampling outcomes adds much coverage in this regard.

For reference, the existing `test_qsim_run_vs_cirq_run` uses a single-outcome circuit to specifically examine the behavior of that outcome. The result is asymmetric (0101) to check for bit-ordering errors, and multiple measurements are extracted to verify that mapping from raw data to keyed results is working.

**Comment by sergeisakov** [comment]
SGTM

**Pull request merged**